backend.go

```diff
@@ -15,12 +18,42 @@ func NewFilesystemBackend(root, tmp string) (storage.Backend, error) {
 		return nil, err
 	}
 
+	storage, err := findAllBackends(fsobj, packs, root)
+	if err != nil {
+		return nil, err
+	}
+
 	return &filesystemBackend{
 		fs:       fsobj,
-		backends: []storage.Storage{fsobj, packs},
+		backends: storage,
 	}, nil
 }
 
+func findAllBackends(mainLoose *fileStorer, mainPacked *pack.Storage, root string) ([]storage.Storage, error) {
+	storage := make([]storage.Storage, 2)
+	storage[0] = mainLoose
+	storage[1] = mainPacked
+	f, err := os.Open(path.Join(root, "info", "alternates"))
+	if err != nil {
+		// No alternates file, no problem.
+		if err != os.ErrNotExist {
+			return storage, nil
+		}
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		storage = append(storage, newFileStorer(scanner.Text(), ""))
```

Comment:
I think that this is how Git does it, though perhaps with some tighter bounds, i.e., that Git will complain specifically if a directory is not a `.git/objects` somewhere inside a repository. Is there anything that we need to do to ensure that we're not opening arbitrary files all over? I don't think so, since I'm pretty sure that this is how Git handles this case, too.

Reply:
I don't think there is anything.  Git repositories aren't supposed to be used in untrusted contexts, so we should be able to assume that the alternates files are in fact valid and reasonably well formed.  With the code the way it is, if we get a bad location, then we'll simply find zero loose objects and zero pack files in that location, and it will be a no-op.